### ISSUE TITLE
SubDimensions + mpi full mode issue

### ISSUE THREAD

**Issue by tjb900** [devitocodes/devito#1309]
AFAICT, this is something to do with the boundary thicknesses not being taken into account correctly when computing the interior and remainder regions.  So this test reproduces the issue on one rank; no need for mpirun.

Reproducer code

```
from mpi4py import MPI

from devito import Grid, TimeFunction, SubDimension, Eq, Operator, configuration

import logging
logging.basicConfig(level=logging.DEBUG)

configuration["language"] = "openmp"

grid = Grid(shape=(12, 12, 12), comm=MPI.COMM_WORLD)

bc_thickness = 2

internal_subdims = [
                SubDimension.middle(
                    name=d.name + "i",
                    parent=d,
                    thickness_left=bc_thickness,
                    thickness_right=bc_thickness,
                )
                for d in grid.dimensions
            ]

u = TimeFunction(name='u', grid=grid, space_order=4, time_order=1)
eq = Eq(u.forward, u+1+0.0000001*u.dx2)
eq = eq.xreplace({d: di for d, di in zip(grid.dimensions, internal_subdims)})

print(eq)

op = Operator(eq)

u.data[:] = 0

op.apply(time_m=0, time_M=0, xi0_blk0_size=4, yi0_blk0_size=4)

# test point
test_point = [1, bc_thickness, bc_thickness, bc_thickness]
test_point_val = u.data[test_point]

# with mpi=full, gives 0, otherwise 1
print(test_point_val)
```

Output:

```
(devito_contrib_202005) [(devito_contrib..master) devito_contrib]$ DEVITO_MPI=1 OMPI_CC=gcc python full_bug_202005.py
Eq(u(t + dt, xi, yi, zi), u(t, xi, yi, zi) + 1.0e-7*Derivative(u(t, x, y, z), (x, 2)) + 1)
Operator `Kernel` run in 0.01 s
INFO:Devito:Operator `Kernel` run in 0.01 s
1.0
(devito_contrib_202005) [(devito_contrib..master) devito_contrib]$ DEVITO_MPI=full OMPI_CC=gcc python full_bug_202005.py
Eq(u(t + dt, xi, yi, zi), u(t, xi, yi, zi) + 1.0e-7*Derivative(u(t, x, y, z), (x, 2)) + 1)
Operator `Kernel` run in 0.01 s
INFO:Devito:Operator `Kernel` run in 0.01 s
0.0
```

**Comment by rhodrin** [comment]
Thanks for the report and MFE. We've honed in on the culprit now, and are just exploring some possible fixes.

**Comment by FabioLuporini** [comment]
closing as fixed by #1320
